### PR TITLE
Add option to prefix sitemap

### PR DESCRIPTION
**Pull request by ktym4a**
## Changes

Google Serach Console sometimes not work by `sitemap-*.xml`, so add an option to change the file name.
https://github.com/ekalinin/sitemap.js/pull/395

Follow this one with `Want to customize that?` section.
https://github.com/ekalinin/sitemap.js?tab=readme-ov-file#create-sitemap-and-index-files-from-one-large-list

## Testing

Add prefix.test.js.
Test the content is the same with and without the prefix option.
Test the link is in `*-index.xml` correct.

## Docs

Docs need to be updated, because there are more options

### CODE REVIEW

**Comment by ktym4a** [comment]
I'm sure that `prefix` need to be checked for correct input.
Should I check in `schema.ts` and if input is not correct, pass `undefined` and set `sitemap-`? (I mean I set `sitemap-` for default)
https://github.com/withastro/astro/blob/b596506a242d05ffc112af2aaa008f29aa6a7f9a/packages/integrations/sitemap/src/index.ts#L83

and Is it okay that I use `/^[a-zA-Z\-_]+$/` for check characters?

also Do you think `prefix` is a good name for this option?

**Comment by ematipico** [comment]
> Google Serach Console sometimes not work by sitemap-*.xml

Do you have a link that explains this?

**Comment by ktym4a** [comment]
@ematipico 
I have not seen something from official, but the same issue is listed below.
https://beiznotes.org/how-to-avoid-the-0-discovered-urls-bug-in-search-console/
https://github.com/ekalinin/sitemap.js/pull/395

and I try like this to see, if filename is not `sitemap-${i}.xml` then work.
https://ktym4a.me/sitemap-index.xml (Discovered pages: 0) copy to https://ktym4a.me/sitemap-aa.xml  and change url sitemap-0.xml to sitemap-a.xml (Discovered pages: 6)
https://ktym4a.me/sitemap-0.xml (Couldn't fetch) copy to https://ktym4a.me/sitemap-a.xml (Discovered pages: 6) 

Screenshot 2024-01-27 at 22 14 41


But, this bug is really strange, that I have two static sites with astro and one works correctly with sitemap-index.xml and sitemap-0.xml.

**Comment by ktym4a** [comment]
also this one
https://www.seroundtable.com/google-fetch-process-sitemap-file-31042.html

**Comment by ktym4a** [comment]
I tested again now.

`default`
https://ktym4a.me/sitemap-index.xml (Discovered pages: 0)
https://ktym4a.me/sitemap-0.xml (Couldn't fetch)

`change ktym4a-*`
https://ktym4a.me/ktym4a-index.xml (Discovered pages: 0)
https://ktym4a.me/ktym4a-0.xml (Couldn't fetch)

`change sitemap_*`
https://ktym4a.me/sitemap_index.xml (Discovered pages: 6)
https://ktym4a.me/sitemap_0.xml (Discovered pages: 6)

Screenshot 2024-01-28 at 16 46 41

File: .changeset/eighty-falcons-tease.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@astrojs/sitemap": patch
```
Comment (ematipico) [inline .changeset/eighty-falcons-tease.md:2]: As we are adding a new option, this should be a `minor` and not a `patch`

File: packages/integrations/sitemap/src/index.ts [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
+import path, { resolve } from 'node:path';
 import { fileURLToPath } from 'node:url';
 import type { EnumChangefreq, LinkItem as LinkItemBase, SitemapItemLoose } from 'sitemap';
-import { simpleSitemapAndIndex } from 'sitemap';
+import { SitemapAndIndexStream, SitemapStream, streamToPromise } from 'sitemap';
 import { ZodError } from 'zod';
 
 import { generateSitemap } from './generate-sitemap.js';
 import { validateOptions } from './validate-options.js';
+import { createWriteStream } from 'fs';
```
Comment (ematipico) [inline packages/integrations/sitemap/src/index.ts:10]:
```suggestion
import { createWriteStream } from 'node:fs';
```

**Review by ematipico — APPROVED** [review]
The changes look good. I will ask docs to review the changelog. Usually, when we add a new feature, we also want to explain that with some code snippet.

**Comment by sarah11918** [comment]
Hi @ktym4a -- I'm here to help with docs!

I'll make a suggestion for the changeset message for the kind of message we would expect here, and then you can fix it as necessary. 

There should also be a PR made to the docs repo to update this page/section that matches the other configuration options: https://docs.astro.build/en/guides/integrations-guide/sitemap/#configuration

When you make that PR, please fill out the template for adding a new feature to Astro v4.4 and include a link to this PR so we keep them organized together. Don't worry if that PR will need editing, I'll be there to help too! Just try to get it started similar to the other configuration options there. It doesn't need to be fancy, just follow the existing ones as a guide.

File: .changeset/eighty-falcons-tease.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"@astrojs/sitemap": minor
+---
+
+Added option to prefix sitemap (default is still sitemap-\*)
```
Comment (sarah11918) [inline .changeset/eighty-falcons-tease.md:5]:
``````suggestion
Adds a new configuration option `prefix` that allows you to change the default `sitemap-*.xml` file name.

By default, running `astro build` creates both `sitemap-index.xml` and `sitemap-0.xml` in your output directory. 

To change the names of these files (e.g. to `astrosite-index.xml` and `astrosite-0.xml`), set the `prefix` option in your `sitemap` integration configuration:

```
import { defineConfig } from 'astro/config';
import sitemap from '@astrojs/sitemap';

export default defineConfig({
  site: 'https://example.com',
  integrations: [
    sitemap({
      prefix: 'astrosite-',
    }),
  ],
});
```
``````

**Review by sarah11918 — COMMENTED** [review]
Hi @ktym4a !

This is an example of what we would normally have here! Since you are the expert in the feature, I will need you to make this actually correct and fix anything. But, I think it would be nice to have this level of detail here.

You could also maybe add a sentence at the end like:

> This option is useful when...

But that's up to you!

File: .changeset/eighty-falcons-tease.md (hunk 1 above)
Comment (ktym4a) [reply .changeset/eighty-falcons-tease.md]: I changed it.

File: .changeset/eighty-falcons-tease.md (hunk 3 above)
Comment (ktym4a) [reply .changeset/eighty-falcons-tease.md]: I updated! thanks for your support.

**Comment by ktym4a** [comment]
@ematipico
I did not explain well enough, sorry.
The reason I told it is because if someone sets like this, it will be created outside of the output directory.
```ts
export default defineConfig({
  integrations: [sitemap({
    prefix: '../a-'
  })]
})
```

**Comment by ematipico** [comment]
Oh, I understand now. Yeah, it's a very edge case. If you're up to the task, feel free to add some level of validation, but I wouldn't worry too much if you don't want to implement it. I think it's fine like this too. It's a very edge case if a user decides to use this "trick", they can report it and just putting a plain string without paths-like characters.

File: .changeset/eighty-falcons-tease.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  site: 'https://example.com',
+  integrations: [
+    sitemap({
+      prefix: 'astrosite-',
+    }),
+  ],
+});
+```
+
+This option is useful when you submit a sitemap to Google Search Console and the Status becomes `Cloudn't fetch`
```
Comment (sarah11918) [inline .changeset/eighty-falcons-tease.md:24]:
```suggestion
This option is useful when Google Search Console is unable to fetch your default sitemap files, but can read renamed files.
```

Comment (ktym4a) [reply .changeset/eighty-falcons-tease.md]: It feels perfect to me! Thank you.

**Review by sarah11918 — COMMENTED** [review]
Approving this PR on behalf of docs! Thank you for this helpful contribution @ktym4a :raised_hands: 

(And I can see another PR in the docs repo, so I'll get to that one soon)

**Comment by sarah11918** [comment]
I've already approved here, but I'll also note so it's clear here that the accompanying Docs PR has been approved, so docs is ready whenever you decide to release a new version of `@astrojs/sitemap`!

**Pull request merged**